File: academic-system-python/academic_system/security/session.py
from typing import Optional

from ..model.user import User


class Session:
    _instance: Optional["Session"] = None

    def __init__(self):
        self._authenticated_user: Optional[User] = None

    @classmethod
    def get_instance(cls) -> "Session":
        if cls._instance is None:
            cls._instance = Session()
        return cls._instance

    @classmethod
    def reset(cls) -> None:
        cls._instance = None

    def login(self, user: User) -> None:
        self._authenticated_user = user

    def logout(self) -> None:
        self._authenticated_user = None

    def get_authenticated_user(self) -> Optional[User]:
        return self._authenticated_user

    def is_authenticated(self) -> bool:
        return self._authenticated_user is not None

File: academic-system-java/src/main/java/org/example/academic/system/security/Session.java
package org.example.academic.system.security;

import org.example.academic.system.model.User;

public class Session {
    private static Session instance;
    private User authenticatedUser;

    private Session() {
    }

    public static Session getInstance() {
        if (instance == null) {
            instance = new Session();
        }
        return instance;
    }

    public void login(User user) {
        this.authenticatedUser = user;
    }

    public User getAuthenticatedUser() {
        return authenticatedUser;
    }

    public void logout() {
        this.authenticatedUser = null;
    }

    public boolean isAuthenticated() {
        return authenticatedUser != null;
    }
}
